**Материал по лекции:**

In [6]:
df.tail()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('sample_data/california_housing_train.csv')

In [5]:
df.head(n=7)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   
5    -114.58     33.63                29.0       1387.0           236.0   
6    -114.58     33.61                25.0       2907.0           680.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0  
5       671.0       239.0         3.3438             74000.0  
6      1841.0       633.0         2.6768             82400.0

In [7]:
df.shape

(17000, 9)

In [8]:
df.isnull()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0          False     False               False        False           False   
1          False     False               False        False           False   
2          False     False               False        False           False   
3          False     False               False        False           False   
4          False     False               False        False           False   
...          ...       ...                 ...          ...             ...   
16995      False     False               False        False           False   
16996      False     False               False        False           False   
16997      False     False               False        False           False   
16998      False     False               False        False           False   
16999      False     False               False        False           False   

       population  households  median_income  median_house_value  
0           False       False          False               False  
1           False       False          False               False  
2           False       False          False               False  
3           False       False          False               False  
4           False       False          False               False  
...           ...         ...            ...                 ...  
16995       False       False          False               False  
16996       False       False          False               False  
16997       False       False          False               False  
16998       False       False          False               False  
16999       False       False          False               False  

[17000 rows x 9 columns]

In [9]:
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
dtype: int64

In [10]:
df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
dtype: object

In [11]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value'],
      dtype='object')

In [12]:
df['latitude']

0        34.19
1        34.40
2        33.69
3        33.64
4        33.57
         ...  
16995    40.58
16996    40.69
16997    41.84
16998    41.80
16999    40.54
Name: latitude, Length: 17000, dtype: float64

In [13]:
df[['latitude', 'population']]

latitude  population
0         34.19      1015.0
1         34.40      1129.0
2         33.69       333.0
3         33.64       515.0
4         33.57       624.0
...         ...         ...
16995     40.58       907.0
16996     40.69      1194.0
16997     41.84      1244.0
16998     41.80      1298.0
16999     40.54       806.0

[17000 rows x 2 columns]

In [16]:
df[(df['housing_median_age'] > 20) & (df['total_rooms'] > 2000)]

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
6        -114.58     33.61                25.0       2907.0           680.0   
8        -114.59     33.61                34.0       4789.0          1175.0   
13       -114.61     34.83                31.0       2478.0           464.0   
42       -115.49     32.67                25.0       2322.0           573.0   
45       -115.50     32.67                35.0       2159.0           492.0   
...          ...       ...                 ...          ...             ...   
16986    -124.19     40.73                21.0       5694.0          1056.0   
16990    -124.22     41.73                28.0       3003.0           699.0   
16993    -124.23     40.54                52.0       2694.0           453.0   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   

       population  households  median_income  median_house_value  
6          1841.0       633.0         2.6768             82400.0  
8          3134.0      1056.0         2.1782             58400.0  
13         1346.0       479.0         3.2120             70400.0  
42         2185.0       602.0         1.3750             70100.0  
45         1694.0       475.0         2.1776             75500.0  
...           ...         ...            ...                 ...  
16986      2907.0       972.0         3.5363             90100.0  
16990      1530.0       653.0         1.7038             78300.0  
16993      1152.0       435.0         3.0806            106700.0  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  

[5624 rows x 9 columns]

In [18]:
df['population'].max()

35682.0

In [19]:
df['population'].min()

3.0

In [20]:
df['population'].mean()

1429.5739411764705

In [21]:
df['population'].sum()

24302757.0

In [22]:
df[['population', 'total_rooms']].median()

population     1167.0
total_rooms    2127.0
dtype: float64

In [23]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  17000.000000  17000.000000        17000.000000  17000.000000   
mean    -119.562108     35.625225           28.589353   2643.664412   
std        2.005166      2.137340           12.586937   2179.947071   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.790000     33.930000           18.000000   1462.000000   
50%     -118.490000     34.250000           29.000000   2127.000000   
75%     -118.000000     37.720000           37.000000   3151.250000   
max     -114.310000     41.950000           52.000000  37937.000000   

       total_bedrooms    population    households  median_income  \
count    17000.000000  17000.000000  17000.000000   17000.000000   
mean       539.410824   1429.573941    501.221941       3.883578   
std        421.499452   1147.852959    384.520841       1.908157   
min          1.000000      3.000000      1.000000       0.499900   
25%        297.000000    790.000000    282.000000       2.566375   
50%        434.000000   1167.000000    409.000000       3.544600   
75%        648.250000   1721.000000    605.250000       4.767000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        17000.000000  
mean        207300.912353  
std         115983.764387  
min          14999.000000  
25%         119400.000000  
50%         180400.000000  
75%         265000.000000  
max         500001.000000

# **Домашнее задание к уроку 9**

**********
**ЗАДАЧА 1.**
Работать с файлом california_housing_train.csv, который находится в папке
sample_data. Определить среднюю стоимость дома, где кол-во людей от 0 до 500 (population)

In [27]:
# Выборка всех строк, где популяция от 0 до 500
df_popul0_500 = df[(df['population'] > 0) & (df['population'] < 500)]
df_popul0_500

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2        -114.56     33.69                17.0        720.0           174.0   
7        -114.59     34.83                41.0        812.0           168.0   
17       -114.65     32.79                21.0         44.0            33.0   
19       -114.67     33.92                17.0         97.0            24.0   
21       -114.73     33.43                24.0        796.0           243.0   
...          ...       ...                 ...          ...             ...   
16944    -124.14     40.67                23.0        580.0           117.0   
16951    -124.15     40.81                50.0        340.0            74.0   
16969    -124.17     40.80                52.0        661.0           316.0   
16982    -124.18     40.62                35.0        952.0           178.0   
16994    -124.25     40.28                32.0       1430.0           419.0   

       population  households  median_income  median_house_value  
2           333.0       117.0         1.6509             85700.0  
7           375.0       158.0         1.7083             48500.0  
17           64.0        27.0         0.8571             25000.0  
19           29.0        15.0         1.2656             27500.0  
21          227.0       139.0         0.8964             59200.0  
...           ...         ...            ...                 ...  
16944       320.0       109.0         4.2054            130600.0  
16951       235.0        83.0         1.7500             67500.0  
16969       392.0       244.0         0.9570             60000.0  
16982       480.0       179.0         3.0536            107000.0  
16994       434.0       187.0         1.9417             76100.0  

[1601 rows x 9 columns]

In [28]:
# Столбец со средней стоимостью дома из выборки df_popul0_500
df_popul0_500['median_house_value']

2         85700.0
7         48500.0
17        25000.0
19        27500.0
21        59200.0
           ...   
16944    130600.0
16951     67500.0
16969     60000.0
16982    107000.0
16994     76100.0
Name: median_house_value, Length: 1601, dtype: float64

In [29]:
# Средняя стоимость дома из выборки выше
df_popul0_500['median_house_value'].mean()

206683.83635227982

In [30]:
# Если попробовать получить результат одной строкой
df[(df['population'] > 0) & (df['population'] < 500)]['median_house_value'].mean()

206683.83635227982

**********
**ЗАДАЧА 2.**
Узнать какая максимальная households в зоне минимального значения population

In [48]:
# Выборка двух столбцов
df_populationHouseholds = df[['population', 'households']]
df_populationHouseholds

population  households
0          1015.0       472.0
1          1129.0       463.0
2           333.0       117.0
3           515.0       226.0
4           624.0       262.0
...           ...         ...
16995       907.0       369.0
16996      1194.0       465.0
16997      1244.0       456.0
16998      1298.0       478.0
16999       806.0       270.0

[17000 rows x 2 columns]

In [50]:
# Статистика по ним
df_populationHouseholds.describe()

population    households
count  17000.000000  17000.000000
mean    1429.573941    501.221941
std     1147.852959    384.520841
min        3.000000      1.000000
25%      790.000000    282.000000
50%     1167.000000    409.000000
75%     1721.000000    605.250000
max    35682.000000   6082.000000

In [73]:
# Минимальная и максимальная популяции
populMin = df_populationHouseholds['population'].min()
populMax = df_populationHouseholds['population'].max()
populMin, populMax

(3.0, 35682.0)

Дадим оценку термину "зона минимального значения population".
Видим, что min = 3, max = 35682.
Разберем случай, когда популяция < 1000, а потом в цикле прогоним несколько вариантов

In [93]:
# Выборка, где популяция < 1000
min = 1000
df_min = df[df['population'] < min][['population', 'households']]
df_min

population  households
2           333.0       117.0
3           515.0       226.0
4           624.0       262.0
5           671.0       239.0
7           375.0       158.0
...           ...         ...
16989       620.0       268.0
16992       544.0       172.0
16994       434.0       187.0
16995       907.0       369.0
16999       806.0       270.0

[6689 rows x 2 columns]

In [94]:
# В таком случае максимальная households
householdsMax = df_min['households'].max()
householdsMax

723.0

Теперь циклом разберем много вариантов

In [95]:
householdsMaxDict = {}
householdsMaxDict[min] = householdsMax
print(f"В зоне значений population до {min}, максимальная households = {householdsMaxDict[min]}")
popul = 2500
while (popul < populMax):
  df_popul = df[df['population'] < popul][['population', 'households']]
  householdsMaxDict[popul] = df_popul['households'].max()
  print(f"В зоне значений population до {popul}, максимальная households = {householdsMaxDict[popul]}")
  popul += 2500

В зоне значений population до 1000, максимальная households = 723.0
В зоне значений population до 2500, максимальная households = 1605.0
В зоне значений population до 5000, максимальная households = 2565.0
В зоне значений population до 7500, максимальная households = 4204.0
В зоне значений population до 10000, максимальная households = 4616.0
В зоне значений population до 12500, максимальная households = 4616.0
В зоне значений population до 15000, максимальная households = 4616.0
В зоне значений population до 17500, максимальная households = 5189.0
В зоне значений population до 20000, максимальная households = 5189.0
В зоне значений population до 22500, максимальная households = 5189.0
В зоне значений population до 25000, максимальная households = 5189.0
В зоне значений population до 27500, максимальная households = 5189.0
В зоне значений population до 30000, максимальная households = 6082.0
В зоне значений population до 32500, максимальная households = 6082.0
В зоне значений populatio